### PR TITLE
Elixir: Add telemetry support

### PR DESCRIPTION
**Pull request by halostatue**
- Telemetry will only be enabled if `telemetry` is in the hosting application's dependencies. It may also be explicitly disabled:

  ```elixir
  config :app_identity, AppIdentity.Telemetry, enabled: false
  ```

- Telemetry is emitted as internally-implemented spans for `AppIdentity.generate_proof/2`, `AppIdentity.verify_proof/3`, and `AppIdentity.Plug`.

- Testing has been improved as the tests can now *correctly* verify against the external API (`AppIdentity`) rather than the internal API (`AppIdentity.Internal`), but verifies the internal implementation using the telemetry output.

### CODE REVIEW

File: elixir/lib/app_identity.ex [hunk 1]
```diff
@@ -185,6 +208,12 @@ defmodule AppIdentity do
 
       iex> AppIdentity.generate_proof!(v2(), disallowed: [1, 2])
       ** (AppIdentity.AppIdentityError) Error generating proof
+
+  ### Telemetry
+
+  When telemetry is enabled, `generate_proof/2` will emit `[:app_identity,
```
Comment (mikestok) [inline elixir/lib/app_identity.ex:214]: Should this be `generate_proof!/2`?

File: elixir/lib/app_identity.ex [hunk 2]
```diff
@@ -279,6 +313,12 @@ defmodule AppIdentity do
   ```elixir
   AppIdentity.verify_proof(proof, &IdentityApplications.get!(&1.id))
   ```
+
+  ### Telemetry
+
+  When telemetry is enabled, `verified_proof/3` will emit `[:app_identity,
```
Comment (mikestok) [inline elixir/lib/app_identity.ex:319]: Should this be `verify_proof!/3`

File: elixir/lib/app_identity/telemetry.ex [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                optional(:verified) => term()
+              }
+
+  # Span context returned from `start_span/2` and should be passed to
+  # `stop_span/2`.
+  @typep span_context :: {telemetry_method, start_time :: term()}
+
+  # Start a telemetry span.
+  @doc false
+  @spec start_span(telemetry_method, metadata :: term) ::
```
Comment (mikestok) [inline elixir/lib/app_identity/telemetry.ex:155]: Method? Should that be function?

**Review by mikestok — DISMISSED** [review]

File: elixir/lib/app_identity.ex [hunk 4]
```diff
@@ -185,6 +208,12 @@ defmodule AppIdentity do
 
       iex> AppIdentity.generate_proof!(v2(), disallowed: [1, 2])
       ** (AppIdentity.AppIdentityError) Error generating proof
+
+  ### Telemetry
+
+  When telemetry is enabled, `generate_proof/2` will emit `[:app_identity,
+  :generate_proof, :start]` and `[:app_identity, :generate_proof, :stop]`
```
[the remark this answers is not in the document]
Comment (halostatue) [reply elixir/lib/app_identity.ex:215]: `generate_proof!/2` is how `generate_proof/2` is called throwing an exception if not. The implementation is still `generate_proof/2`, behind the scenes, so the telemetry event is the same.

**Review by mikestok — APPROVED** [review]

**Pull request merged**